v2/cmd/asoctl/pkg/importresources/resource_importer.go

```diff
@@ -225,68 +258,78 @@ func (ri *ResourceImporter) importWorker(
 			completed <- imported
 		}
 	}
-
-	done <- struct{}{}
 }
 
-func (ri *ResourceImporter) collateResults(
+// startCollationOfResults starts a goroutine that will collate the results of imports.
+// completed is the channel of completed imports to drain.
+// candidates is the channel that receives any new resources to import.
+// watchdog is updated to track the number of inflight operations.
+// waitgroup is used to track running goroutines.
+// report is the report to write to.
+func (ri *ResourceImporter) startCollationOfResults(
 	completed <-chan ImportResourceResult, // completed imports for us to collate
 	candidates chan<- ImportableResource, // additional candidates for importing
-	progress importreporter.Interface, // importreporter tracking
+	watchdog *watchdog,
+	waitgroup *conc.WaitGroup,
 	report *resourceImportReport, // report to write to
-	done chan<- struct{}, // channel to signal completion
 ) {
-	for importResult := range completed {
-		rsrc := importResult.resource
-		gk := rsrc.GroupKind()
-
-		// Enqueue any child resources we found; these will be marked as completed when we collate their results
-		progress.AddPending(len(importResult.pending))
-		for _, p := range importResult.pending {
-			candidates <- p
-		}
-
-		ri.log.Info(
-			"Imported",
-			"kind", gk,
-			"name", rsrc.Name())
+	waitgroup.Go(func() {
+		for importResult := range completed {
+			rsrc := importResult.resource
+			gk := rsrc.GroupKind()
+
+			// Enqueue any child resources we found; these will be marked as completed when we collate their results
+			ri.reporter.AddPending(len(importResult.pending))
+			for _, p := range importResult.pending {
+				candidates <- p
+			}
 
-		report.AddSuccessfulImport(gk)
-		ri.imported[rsrc.ID()] = rsrc
+			ri.log.Info(
+				"Imported",
+				"kind", gk,
+				"name", rsrc.Name())
 
-		// Flag the main resource as complete
-		// We do this after everything else because it might indicate we're finished
-		progress.Completed(1)
-	}
+			report.AddSuccessfulImport(gk)
+			ri.imported[rsrc.ID()] = rsrc
 
-	done <- struct{}{}
+			// Flag the main resource as complete
+			// We do this after everything else because it might indicate we're finished
+			ri.reporter.Completed(1)
+			watchdog.stopped()
```

Comment:
I'd put these into a method on ResourceImporter and then call it from both the success and error workers. Tying them together makes it harder to forget to do it in one place.

Reply:
Done.